Desktop Access without Active Directory

alexfornuto opened the pull request:
Resolves #18996 

Updates the position and title of existing desktop access docs, with a new Getting Started guide that covers Desktop access without Active Directory.

Commit:
rename manual to active directory manual

Commit:
rename getting started to active directory

Commit:
update titles and descriptions

Commit:
first draft of non-AD getting started

Note: I have images but did not yet include them in the commit history. No need to bloat the repo if those steps will be invalidated by a new installer

alexfornuto commented on docs/pages/desktop-access/active-directory.mdx:
```diff
@@ -0,0 +1,213 @@
+---
```
I can't figure out why GitHub shows this as a new and not renamed file, when the [commit](https://github.com/gravitational/teleport/pull/19311/commits/9a983a3592a1dbb452d02689ea28b0096509c015) shows it as a rename.

zmb3 commented on docs/config.json:
```diff
@@ -900,8 +900,12 @@
         {
-          "title": "Manual Setup",
-          "slug": "/desktop-access/manual-setup/"
+          "title": "Active Directory",
```
I'm not sure this title is descriptive enough now. Maybe `Set up with Active Directory`?

alexfornuto replied:
Note that this is the title displayed in the sidebar, not the title of the document.

zmb3 replied:
Ah, no concerns in that case then.

zmb3 commented on docs/pages/desktop-access/getting-started.mdx:
```diff
@@ -1,213 +1,134 @@
-description: Automatically add Teleport Desktop Access to your infrastructure
+description: Access Windows systems through Teleport Desktop Access
 ---
 
```
How about a note at the top to help orient the reader and ensure they're on the right page? Something like:

```
Using Active Directory?

This guide sets up Teleport Desktop Access for standalone hosts that are not
joined to an Active Directory domain. To integrate Teleport with AD, see our
Active Directory guide.
```

alexfornuto replied:
Yes, I still need to write the intro to the doc. This is a draft PR, staged as such while I make sure the technical details are correct as they are confirmed.

zmb3 commented on docs/pages/desktop-access/getting-started.mdx:
```diff
@@ -1,213 +1,134 @@
+- An existing Teleport cluster with one of the following versions:
 
-</Details>
+  **Open Source or Enterprise:** version 8.0 or newer
```
For non-AD this will be 12.0 or newer.

alexfornuto replied:
Yup, still need to update a lot in this. Draft PR.

zmb3 commented on docs/pages/desktop-access/getting-started.mdx:
```diff
@@ -1,213 +1,134 @@
+Reboot the Windows system after installing the DLL.
 
-Once you've saved `/etc/teleport.yaml`, start Teleport:
+### Disable network level authenticate for RDP
```
```suggestion
### Disable Network Level Authentication for RDP
```

Commit:
update draft

alexfornuto commented on docs/pages/desktop-access/getting-started.mdx:
```diff
@@ -1,214 +1,171 @@
 
-Copy and paste the provided command into a Windows PowerShell window to download
-and run the 'configure Active Directory' script:
+From the Windows system, download the [Teleport service installer](#LinkToCome) {/*Rename as needed */}
```
The link on this line needs to be replaced with the final name and download location of the installer.

Commit:
update links for new paths

alexfornuto commented:
One thing to consider: Do we want to change the naming in these docs from "Desktop Access" to "WIndows Protocol"?

zmb3 commented on docs/pages/desktop-access/getting-started.mdx:
````diff
@@ -1,214 +1,171 @@
-```yaml
-version: v3
+```yml
+version: v6
````
```suggestion
version: v3
```

Roles were bumped to v6, but the latest config version is still v3.

zmb3 commented on docs/pages/desktop-access/getting-started.mdx:
```diff
@@ -1,214 +1,171 @@
 teleport:
-  auth_token: 2239...c5b21
-  proxy_server: teleport.example.com:443
+  nodename: example.teleport.com
```
This is also missing the auth/proxy server address.

alexfornuto replied:
Added.

zmb3 commented on docs/pages/desktop-access/getting-started.mdx:
```diff
@@ -1,214 +1,171 @@
+  nodename: example.teleport.com
+windows_desktop_service:
+  enabled: yes
+  listen_addr: 0.0.0.0:3028
```
We should probably prefer to use tunneling like we do in the other guide. (Connect to proxy on 443/3080)

alexfornuto replied:
I removed the key based on the other guide.

zmb3 commented on docs/pages/desktop-access/getting-started.mdx:
```diff
@@ -1,214 +1,171 @@
+windows_desktop_service:
+  enabled: yes
+  listen_addr: 0.0.0.0:3028
+  non_ad_hosts:
```
Let's add this new field to the config reference as well.

alexfornuto replied:
Done.

zmb3 commented on docs/pages/desktop-access/getting-started.mdx:
```diff
@@ -1,214 +1,171 @@
+Ensure that each Teleport user is only assigned Windows logins that they should
+be allowed to access.
+
+Usernames shared between domain and local users will create login conflicts.
```
Since this guide is about non-AD, I don't think domain users are are play. Should we remove this?

zmb3 commented on docs/pages/desktop-access/getting-started.mdx:
```diff
@@ -1,214 +1,171 @@
-Copy and paste the provided command into a Windows PowerShell window to download
-and run the 'configure Active Directory' script:
+From the Windows system, download the [Teleport service installer](#LinkToCome) {/*Rename as needed */}
+and run it. When prompted, select the Teleport certificate file from the previous step.
```
There's also a headless install option that might be worth documenting here.

alexfornuto replied:
I'd be happy to document it once I learn about it myself.

zmb3 replied:
You should be able to invoke the same installer from the command line with a `--help` flag to see the options.

zmb3 replied:
@probakowski please advise ☝️

probakowski replied:
It turns out it will show correct output if you run it (with `--help`) from elevated prompt. If you do it from normal prompt it will show UAC screen and then start new elevated console which will close immediately as the program exits (same thing will happen if you create shortcut with `--help` parameter and run it by double click)

alexfornuto replied:
I've added a tip below.

Commit:
Update based on review feedback

Commit:
update config reference

ptgott commented on docs/pages/desktop-access/active-directory-manual.mdx:
```diff
@@ -1,6 +1,6 @@
+title: Desktop Access with Active Directory (Manual)
+description: Manually Connect Teleport to an Active Directory domain
 videoBanner: YvMqgcq0MTQ
 ---
```
Should this link to the other Active Directory guide so the user knows when to use each one?

ptgott commented on docs/pages/desktop-access/active-directory.mdx:
```diff
@@ -0,0 +1,216 @@
+  [Manual Setup](active-directory-manual.mdx) guide for Desktop Access.</ScopedBlock>
+- An Active Directory domain, configured for LDAPS (Teleport requires an
+  encrypted LDAP connection). Typically this means installing
+  [AD CS](https://learn.microsoft.com/en-us/windows-server/identity/ad-cs/)
```
```suggestion
  [AD CS](https://learn.microsoft.com/en-us/windows-server/identity/ad-cs/).
```

ptgott commented on docs/pages/desktop-access/active-directory.mdx:
```diff
@@ -0,0 +1,216 @@
+In your web browser, access the teleport Web UI at <ScopedBlock scope={["oss", "enterprise"]}>
+`teleport.example.com`</ScopedBlock><ScopedBlock scope={["cloud"]}>
+`mytennant.teleport.sh`</ScopedBlock>. Click on your user name at the top right
+and select **Manage Access**, Select "Desktop" from the main menu, then **NEXT**:
```
```suggestion
and select **Manage Access**, Select "Desktop" from the main menu, then **NEXT**.
```

ptgott commented on docs/pages/desktop-access/active-directory.mdx:
```diff
@@ -0,0 +1,216 @@
+Once the server is restarted from one or both command executions, click **NEXT**.
+
+Copy and paste the provided command into a Windows PowerShell window to download
+and run the 'configure Active Directory' script:
```
```suggestion
and run the "configure Active Directory" script:
```

ptgott commented on docs/pages/desktop-access/active-directory.mdx:
````diff
@@ -0,0 +1,216 @@
+```yaml
+version: v3
+teleport:
+  auth_token: 2239...c5b21
````
This renders in two different colors in the code snippet. Should we wrap it in quotes or use a variable?

alexfornuto replied:
Replaced with a token preset var.

ptgott commented on docs/pages/desktop-access/active-directory.mdx:
```diff
@@ -0,0 +1,216 @@
+    domain:   'windows.teleport.example.com'
+    username: 'WIN\svc-teleport'
+    server_name: 'windows-server-hostname'
+    insecure_skip_verify: false
```
We might want to add a note about what this does, since it's not something a user would necessarily clean from the name of the key.

alexfornuto replied:
Noted above.

ptgott commented on docs/pages/desktop-access/active-directory.mdx:
```diff
@@ -0,0 +1,216 @@
+    username: 'WIN\svc-teleport'
+    server_name: 'windows-server-hostname'
+    insecure_skip_verify: false
+    ldap_ca_cert: |
```
Just to double-check, does the automation we mentioned earlier set up TLS on the LDAP server and populate this field automatically?

alexfornuto replied:
What automation are you referring to? If you're talking about the windows installer exe, note that that bin is run on standalone windows machines for non-ad access, separate from this doc.

This doc is just what _was_ the former getting started guide, [renamed](https://github.com/gravitational/teleport/pull/19311/files#r1046457925) to accommodate the convention of non-AD being the default instructions.

alexfornuto replied:
If you're talking about the Wizard in the Teleport UI, then yes; the config generated should include the AD CA.... 90% confident.

ptgott replied:
👍 I suppose I'm really asking, do we need to include instructions for how to supply the value of this field?

zmb3 replied:
It might not hurt, but let's track this separately since it's not related to the non-AD use case.

alexfornuto replied:
Since the text above says:
>paste the configuration provided by the output of
>the previous step. Review and edit the addition as necessarily for your environment

We should include all the content provided in the paste, including this field.

ptgott commented on docs/pages/desktop-access/active-directory.mdx:
```diff
@@ -0,0 +1,216 @@
+        -----END CERTIFICATE-----
+
+  discovery:
+    base_dn: '*'
```
This is another field that I'm not sure a user could guess the meaning of—should we explain what this does?

alexfornuto replied:
pinging @zmb3 & @probakowski as SMEs.

zmb3 replied:
It's covered on the config reference page: https://goteleport.com/docs/desktop-access/reference/configuration/

alexfornuto replied:
And since this is generated from the web UI and meant to be pasted directly from there, I think we can leave it as is, same as above.

ptgott commented on docs/pages/desktop-access/active-directory.mdx:
````diff
@@ -0,0 +1,216 @@
+$ systemctl start teleport.service
+```
+
+Copy the join token to a file on the instance where you will run the Windows Desktop
````
Since we also include this snippet above (with the pre-populated information) should we delete this paragraph and the snippet below?

alexfornuto replied:
Removed.

ptgott commented on docs/pages/desktop-access/active-directory.mdx:
````diff
@@ -0,0 +1,216 @@
+$ ./teleport start --config=/etc/teleport.yaml
+```
+
+Copy the join token to a file on the instance where you will run Windows Desktop
````
Same comment as above re: this paragraph/the snippet below

ptgott commented on docs/pages/desktop-access/getting-started.mdx:
```diff
@@ -1,214 +1,169 @@
+
+If you are running Windows in an AD environment, consider integrating Teleport
+with Active Directory for automatic discovery by reading [Desktop Access with
+Active Directory](./active-directory.mdx). You can also mix Active Directory
```
```suggestion
Active Directory](./active-directory.mdx). You can also mix the Teleport Active Directory
```

ptgott commented on docs/pages/desktop-access/getting-started.mdx:
```diff
@@ -1,214 +1,169 @@
-- An Active Directory domain, configured for LDAPS (Teleport requires an
-  encrypted LDAP connection). Typically this means installing
-  [AD CS](https://learn.microsoft.com/en-us/windows-server/identity/ad-cs/)
+- (!docs/pages/includes/edition-prereqs-tabs.mdx!)
```
Not to bring this issue up again, but I'm not sure what the outer bullet indicates here. What category do the inner bullets belong to that unifies them under the outer bullet? 

If we do want to add the bullets within the `Tabs` component under the same outer bullet, maybe we could include something like:

```markdown
- Make sure you have a running Teleport cluster and client tools:

  (!docs/pages/includes/edition-prereqs-tabs.mdx!)
```

alexfornuto replied:
Adjusted, though we should definitely go back and review this partial's format for better inclusion in this list.

ptgott commented on docs/pages/desktop-access/getting-started.mdx:
```diff
@@ -1,214 +1,169 @@
+  You can reuse an existing server running any other Teleport instance.
 
-### Set up the Teleport Auth and Proxy Services
+- (!docs/pages/includes/tctl.mdx!)
```
Putting a `Details` box in a bullet like this looks sort of awkward to me, like we're trying to work against the design of our docs components. Is there an alternative?

ptgott commented on docs/pages/desktop-access/getting-started.mdx:
```diff
@@ -1,214 +1,169 @@
+- (!docs/pages/includes/tctl.mdx!)
 
-(!docs/pages/includes/database-access/start-auth-proxy.mdx!)
+## Step 1/4. Prepare Windows
```
I think this section would benefit from a short paragraph that explains the architecture of the setup we're working through in the guide. That way, it would be easier for me to form a mental model that I can use to make sense of all the instructions.

alexfornuto replied:
Added below.

ptgott commented on docs/pages/desktop-access/getting-started.mdx:
```diff
@@ -1,214 +1,169 @@
 
-Click **NEXT**.
+On the Linux host where you will run the Desktop Service, install Teleport (if
+you plan on running the Windows protocol on the same host as an existing Teleport
```
```suggestion
you plan on running the Desktop Service on the same host as an existing Teleport
```
For precision, since a protocol isn't what you would be running on the host

alexfornuto replied:
Yea... this was based on some input I got on naming conventions that have since changed...

```suggestion
you plan on running the Windows Service on the same host as an existing Teleport
```

ptgott commented on docs/pages/desktop-access/getting-started.mdx:
```diff
@@ -1,214 +1,169 @@
 
-The PowerShell script will output a Teleport configuration block. Copy this
-block to a temporary location. Click **Next**.
+(!docs/pages/includes/install-linux.mdx!)
```
This partial has its own intro paragraph that conflicts with the one in this guide:

<img width="907" alt="image" src="https://user-images.githubusercontent.com/25250668/214699702-e1cb3681-3871-4879-b872-69825feff599.png">

alexfornuto replied:
Imported a commit from another PR that makes the partial more compatible with preceding content from the docs it's called in.

ptgott commented on docs/pages/desktop-access/getting-started.mdx:
```diff
@@ -1,214 +1,169 @@
-On the Linux host you installed Teleport to run as the Desktop Access connector,
-edit `/etc/teleport.yaml` and paste the configuration provided by the output of
-the previous step. Review and edit the addition as necessarily for your environment:
+Define the Desktop Access service and list the Windows systems configured for
```
```suggestion
Enable the Desktop Service and list the Windows systems configured for
```

ptgott commented on docs/pages/desktop-access/getting-started.mdx:
```diff
@@ -1,214 +1,169 @@
+
+See the [RBAC page](./rbac.mdx) for more information about setting up
+Windows Desktop Access permissions.
 
-Teleport will discover available Windows desktops in the domain, and list them
-under **Desktops**. Click **Finish** then **BROWSE EXISTING RESOURCES** to see them:
+See the [Access Controls Getting Started](../access-controls/getting-started.mdx#step-13-add-local-users-with-preset-roles)
+guide for instructions on how to create or update a user with a given role.
```
I think these should go in a "Next steps" section since they're not required to finish this guide.

alexfornuto commented on docs/pages/desktop-access/getting-started.mdx:
```diff
@@ -1,214 +1,169 @@
 ---
 title: Getting Started with Desktop Access
```
```suggestion
title: Getting Started with Windows Access
```

zmb3 replied:
Desktop Access is fine. Let's keep it.

alexfornuto commented on docs/pages/desktop-access/getting-started.mdx:
```diff
@@ -1,214 +1,169 @@
 title: Getting Started with Desktop Access
-description: Automatically add Teleport Desktop Access to your infrastructure
-videoBanner: v36tlB8D5kk
+description: Access Windows systems through Teleport Desktop Access
```
```suggestion
description: Access Windows systems through Teleport using passwordless authentication.
```

Commit:
post-review edits

Commit:
Apply suggestions from code review

Co-authored-by: Paul Gottschling <paul.gottschling@goteleport.com>

Commit:
note non-AD as enterprise feature

Commit:
expand manual AD process adminition

ptgott commented on docs/pages/desktop-access/getting-started.mdx:
```diff
@@ -1,214 +1,179 @@
 
-(!docs/pages/includes/database-access/start-auth-proxy.mdx!)
+In this section we'll import the Teleport certificate authority (**CA**) file
+to your Windows system, and prepare the it for passwordless access through Teleport.
```
```suggestion
to your Windows system, and prepare it for passwordless access through Teleport.
```

ptgott commented on docs/pages/desktop-access/getting-started.mdx:
```diff
@@ -1,214 +1,179 @@
 
-Click **NEXT**.
+On the Linux host where you will run the Desktop Service, install Teleport (if
+you plan on running the Windows Service on the same host as an existing Teleport
```
```suggestion
you plan on running the Teleport Desktop Service on the same host as an existing Teleport
```
For consistency with the Core Concepts page and other parts of this doc

ptgott reviewed (approved):
Approved with minor suggestions

zmb3 commented on docs/pages/desktop-access/getting-started.mdx:
```diff
@@ -1,214 +1,179 @@
 
-Teleport will discover available Windows desktops in the domain, and list them
-under **Desktops**. Click **Finish** then **BROWSE EXISTING RESOURCES** to see them:
+![Connecting to a Windows desktop from the Web UI](../../img/desktop-access/non-ad-connect.png)
```
I recommend updating this screenshot, as the nav for Teleport 12 has changed, and it would be odd if the docs for a feature that is available in v12+ showed a UI from pre-v12.

alexfornuto replied:
Updated

zmb3 commented on docs/pages/desktop-access/getting-started.mdx:
```diff
@@ -1,214 +1,179 @@
-title: Getting Started with Desktop Access
-description: Automatically add Teleport Desktop Access to your infrastructure
-videoBanner: v36tlB8D5kk
+title: Getting Started with Windows Access
```
```suggestion
title: Getting Started with Desktop Access
```

The pages still appear under a `desktop-access` URL and the menu bar on the left still says "Desktop Access." Let's just stick with desktop access for now.

zmb3 commented on docs/pages/desktop-access/getting-started.mdx:
```diff
@@ -1,214 +1,179 @@
-`teleport.example.com`</ScopedBlock><ScopedBlock scope={["cloud"]}>
-`mytennant.teleport.sh`</ScopedBlock>. Click on your user name at the top right
-and select **Manage Access**, Select "Desktop" from the main menu, then **NEXT**:
+Use `tctl` to export the Teleport root certificate:
```
```suggestion
Use `tctl` to export the Teleport user certificate authority:
```

zmb3 commented on docs/pages/desktop-access/getting-started.mdx:
```diff
@@ -1,214 +1,179 @@
+  non_ad_hosts:
+-    - 192.0.2.156
++    - ip-192-0-2-156.us-east-2.compute.internal
++ host_labels:
```
@probakowski can you confirm that `host_labels` apply to `non_ad_hosts` too? (they should, but I'm not sure if we implemented it)

alexfornuto replied:
I'm 90% sure I confirmed this through testing.

probakowski replied:
Just confirmed that it does apply

zmb3 commented on docs/pages/desktop-access/getting-started.mdx:
```diff
@@ -1,214 +1,179 @@
-edit `/etc/teleport.yaml` and paste the configuration provided by the output of
-the previous step. Review and edit the addition as necessarily for your environment:
+Enable the Desktop Service and list the Windows systems configured for
+access through Teleport in `/etc/teleport.yaml`.
```
Maybe we should be more explicit that inventory management for non-AD is currently the user's responsibility.

Maybe something like:

```
Without Active Directory, Teleport cannot automatically discover your desktops.
To register desktops with Teleport, you can define them in your configuration file
or use Teleport's API to build your own integration.
```

alexfornuto replied:
Updated.

zmb3 commented on docs/pages/desktop-access/getting-started.mdx:
```diff
@@ -1,214 +1,179 @@
-  proxy_server: teleport.example.com:443
+  nodename: example.teleport.com
+  proxy_server: teleport-proxy.example.com:443
+windows_desktop_service:
```
`auth_service` and `proxy_service` (and maybe `ssh_service` too, IIRC) are enabled by default, so even though we're trying to keep this config snippet small, we should probably explicitly disable them here.

alexfornuto replied:
Added.

Commit:
Apply suggestions from code review

Co-authored-by: Zac Bergquist <zac.bergquist@goteleport.com>
Co-authored-by: Paul Gottschling <paul.gottschling@goteleport.com>

Commit:
disable default services for Desktop node

Commit:
document cli access, clarify non-AD listing of hosts

Commit:
add join token for new node

Commit:
update screenshot for new UI

alexfornuto commented on docs/pages/desktop-access/getting-started.mdx:
````diff
@@ -1,214 +1,220 @@
+On your local system authenticated to your Teleport cluster, generate a
+short-lived join token:
 
 ```code
-$ systemctl start teleport.service
+$ tctl tokens add --type=windowsdesktop
+The invite token: (=presets.tokens.first=)
+This token will expire in 60 minutes.
+
+This token enables Desktop Access.  See https://goteleport.com/docs/desktop-access/
+for detailed information on configuring Teleport Desktop Access with this token.
 ```
 
-Copy the join token to a file on the instance where you will run the Windows Desktop
-Service, and then use the following configuration:
+Copy the token to the Linux host where you will run the Desktop service as `/tmp/token`.
````
@ptgott please re-review this section. I realized while responding to other feedback that the instructions lacked a join token for the service node.

ptgott replied:
Added a minor suggestion

zmb3 reviewed (approved):
LGTM. I'll have a better idea about the specific download link in the coming days.

ptgott commented on docs/pages/desktop-access/getting-started.mdx:
```diff
@@ -1,214 +1,220 @@
+## Step 2/4. Install the Teleport Desktop Service
 
-This command assumes the `root` user. Prepend `sudo` otherwise.
+On your local system authenticated to your Teleport cluster, generate a
```
```suggestion
On your local system, authenticate to your Teleport cluster via `tsh` and generate a
```

alexfornuto replied:
Isn't this already covered by `docs/pages/includes/tctl.mdx`, which has the user authenticate?

ptgott replied:
Yeah, that's true. I thought "On your local system authenticated to your Teleport cluster" could have been clearer. Maybe a comma after "system"?

Commit:
copy edit...

Based on https://github.com/gravitational/teleport/pull/19311\#discussion_r1095187559

Commit:
name and link for windows exe

alexfornuto commented:
Just waiting for confirmation of the exe link before merging.

zmb3 committed:
Update download link

Commit:
Merge branch 'master' into afornuto/18996-nonAD-desktop

Commit:
note zip format

Commit:
Merge branch 'master' into afornuto/18996-nonAD-desktop
